openforcefield/typing/engines/smirnoff/parameters.py

```diff
@@ -117,6 +111,304 @@ class NonbondedMethod(Enum):
+        """Convert strings expressions to Quantity and validate the units if requested."""
+        if self._unit is not None:
+            # Convert eventual strings to Quantity objects.
+            value = object_to_quantity(value)
+
+            # Check if units are compatible.
+            try:
+                if not self._unit.is_compatible(value.unit):
+                    raise IncompatibleUnitError(f'{self._name[1:]}={value} should have units of {self._unit}')
+            except AttributeError:
+                # This is not a Quantity object.
+                raise IncompatibleUnitError(f'{self._name[1:]}={value} should have units of {self._unit}')
+        return value
+
+    def _call_converter(self, value, instance):
+        """Correctly calls static and instance converters."""
+        if self._converter is not None:
+            try:
+                # Static function.
+                return self._converter(value)
+            except TypeError:
+                # Instance method.
+                return self._converter(instance, self, value)
+        return value
+
+
+class IndexedParameterAttribute(ParameterAttribute):
+    """The attribute of a parameter with an unspecified number of terms.
+
+    Some parameters can be associated to multiple terms, For example,
+    torsions have parameters such as k1, k2, ..., and ``IndexedParameterAttribute``
+    can be used to encapsulate the sequence of terms.
+
+    The only substantial difference with ``ParameterAttribute`` is that
+    only sequences are supported as values and converters and units are
+    checked on each element of the sequence.
+
+    Currently, the descriptor makes the sequence immutable. This is to
+    avoid that an element of the sequence could be set without being
+    properly validated. In the future, the data could be wrapped in a
+    safe list that would safely allow mutability.
+
+    Parameters
+    ----------
+    default : object, optional
+        When specified, the descriptor makes this attribute optional by
+        attaching a default value to it.
+    unit : simtk.unit.Quantity, optional
+        When specified, only sequences of quantities with compatible units
+        are allowed to be set.
+    converter : callable, optional
+        An optional function that can be used to validate and cast each
+        element of the sequence before setting the attribute.
+
+    Examples
+    --------
+
+    Create an optional indexed attribute with unit of angstrom.
+
+    >>> from simtk import unit
+    >>> class MyParameter:
+    ...     length = IndexedParameterAttribute(default=None, unit=unit.angstrom)
+    ...
+    >>> my_par = MyParameter()
+    >>> my_par.length is None
+    True
+
+    Strings are parsed into Quantity objects.
+
+    >>> my_par.length = ['1 * angstrom', 0.5 * unit.nanometer]
+    >>> my_par.length[0]
+    Quantity(value=1, unit=angstrom)
+
+    Note that ``IndexedParameterAttribute`s are immutable right now. This
```

Comment:
Sounds good! We can decide to implement this here or in a separate PR so that it will be smaller to review. It's up to you. I'll want to add a bunch of tests for the new class as I think I'll have to use something similar also in the new `System`.

Reply:
Let's include it in this PR. This would be so easy to forget (or lose in the issue tracker) that I don't want to have `master` in a temporarily API-broken state.